### DISCUSSION TITLE
Reason behind the Tool design

### DISCUSSION THREAD

**Discussion by chris-sanders** [BrainBlend-AI/atomic-agents#22] [General]
Hi, I'm just finding out about Instructor and atomic-agents sounds like a really interesting way to use it. I love the focus on small composable parts and that it's designed to allow you to swap things out along the way. 

I just have one concern that leaves me a bit hesitant to dive in more. The design of Tools. I'm really surprised that it's designed as a 'copy file into your source' method. I understand that means you can fully customize everything, but it also means any bug or future updates are immediately my responsibility. Even if it's just updating for a new python version it seems to me very problematic to throw packaging to the wind and make every user responsible for the upkeep of tools.

Especially with the ability to monkey patch so easily in python, wouldn't it be better to have a repository of tools, that can be updated and let users fork, patch, or submit new tools if they need to adjust functionality?

I assumed the atomic nature would make community provided agents and tools very useful. But if an agent depends on a tool that is effectively forked doesn't that mean that agents can't be combined together easily?

**Comment by Eigenwise** [comment]
Hello, great question!

I have weighed the pros and cons a lot here, and of course neither solution is perfect, but I believe taking full ownership of the tool is the lesser of two evils for several reasons:
- As you mention, the ability to customize it completely
- Since tools are going to be extremely diverse, they will likely have extremely diverse dependencies, in the end we might end up with 100 tools and 500 dependencies, while you just want to create an agent with the ability to do a web search. 
- The requirements for having a tool in the repository are quite strict, as is their structure. Of course bugs are unavoidable, but the likelihood of a bug being discovered every two seconds is extremely small due to the 100% code coverage requirement. So, in other words, the tools should be as evergreen as possible.

On top of those, many people have slightly different needs or expectations, take for example a search tool:
  - Personally I wanted to combine 3 queries, and just get a single list with all results mixed, but according to relevance
  - Someone else wants the results divided into 3 lists
  - Another person wants a different sorting
  - ...
Soon, the search tool grows in logical complexity, and of course because you have a SearxNG tool, a Serper search tool, a duckduckgo searchtool, ... you expect them to have similar input&output contracts, now you have to have all this complexity in each of these, or you make one giant search tool that does it all...

All of that also potentially influences the actual agents of course, in some cases, since the more free/configurable your input&output is, the harder it will be for AI agents to adhere to the schema, especially if you are working with local/opensource models

I strongly believe that the potential for reducing maintenance complexity both for the framework maintainers and its users is huge!
But, if I am wrong, it's much easier to create a package for the tools, than rip it out of the framework if it is already there!

Now, as per your remark about a community provided agent depending on a tool, this is actually impossible!
See, everything is self-contained with input&output schemas. So this means you could literally swap out the calculator tool for a search tool and use it on an agent that is originally designed for roleplaying, all you do is re-assign the input&output schemas and the system adjusts automatically, not that it will be of much use if you do this, but it won't break.

So, this means in practice that if I create an agent for a search tool, that search tool is going to have a "query" property. But if I want to replace it with another search tool that has "queries" (plural) and a "num_queries" and a "category", none of that is an issue, you just swap the schema.

**Reply by chris-sanders** [reply]
Thanks for the explanation!

**Reply by Eigenwise** [reply]
You're welcome! Thanks again for the question, hope it cleared things up!
